### PR TITLE
[RTL] Connect image update signals.

### LINKED ISSUE #87585: RichTextLabel does not listen to an image's `changed` signal to redraw content
### Tested versions

- Reproducible in: 4.3.dev (https://github.com/godotengine/godot/commit/4b6ad349886288405890b07d4a8da425eb3c97ec), 4.2.1.stable

### System information

Godot v4.2.1.stable unknown - Arch Linux #1 SMP PREEMPT_DYNAMIC Sat, 13 Jan 2024 14:37:14 +0000 - X11 - Vulkan (Forward+) - dedicated AMD Radeon RX 6500 XT (RADV NAVI24) () - AMD Ryzen 5 5600G with Radeon Graphics (12 Threads)

### Issue description

The `RichTextLabel` node allows embedding images through BBCode, which internally loads any `Texture2D` resource on the project path. This can be used with custom `Texture2D` implementations as well.

However, unlike many other nodes such as `TextureRect`, `Sprite2D`, `Sprite3D` and `Button`, the label does not listen to the resource's `changed` signal. Thus, in a situation where the resource has changed and has emitted such signal, the label will not update and redraw it's content, unless forced to do so (by calling `queue_redraw` manually or something else that forces a redraw, like changing the window size).

https://github.com/godotengine/godot/assets/6501975/da50d07f-ffa0-4c7d-9b83-dd6a6d146a97

This can occur in the scenario where a custom `Texture2D` is created which has logic to swap texture files internally, and thus uses `emit_changed` to ensure a redraw is done:

### Steps to reproduce

- Create a custom resource inheriting ``Texture2D``
```gdscript
extends Texture2D
class_name CustomTexture2D

var tex : Texture2D
static var flip := false

func _init():
	choose_texture()

func choose_texture():
	if flip:
		tex = preload("res://icon.png")
	else:
		tex = preload("res://icon_2.png")
	flip = not flip
	emit_changed()

# 2D nodes
func _draw(to_canvas_item, pos, modulate, transpose):
	tex.draw(to_canvas_item, pos, modulate, transpose)

func _draw_rect(to_canvas_item, rect, tile, modulate, transpose):
	tex.draw_rect(to_canvas_item, rect, tile, modulate, transpose)

...
```
- Create a resource and store it in the project files
image
- Create a `RichTextLabel` that uses this new resource
image
- Make the resource swap textures on some other object, like a `Button`:
```gdscript
extends Button

@export var resource : CustomTexture2D

func _on_pressed():
	if resource:
		resource.choose_texture()
```

### Minimal reproduction project (MRP)

[ReproRichTextBug.zip](https://github.com/godotengine/godot/files/14053938/ReproRichTextBug.zip)

### PR DESCRIPTION
**Pull request by bruvzg**
Fixes https://github.com/godotengine/godot/issues/87585

https://github.com/godotengine/godot/assets/7645683/e2320ddd-3aa7-4995-bdc6-7b4e1348b2c1

### CODE REVIEW

**Review by AThousandShips — DISMISSED** [review]
LGTM implementation wise

File: scene/gui/rich_text_label.cpp [hunk 1]
```diff
@@ -3250,6 +3254,9 @@ void RichTextLabel::add_image(const Ref<Texture2D> &p_image, int p_width, int p_
 	item->key = p_key;
 	item->tooltip = p_tooltip;
 
+	item->owner = this;
+	item->image->connect_changed(callable_mp(this, &RichTextLabel::_texture_changed), CONNECT_REFERENCE_COUNTED);
```
Comment (KoBeWi) [inline scene/gui/rich_text_label.cpp:3258]: Why it's not connected to `queue_redraw()` directly?

**Comment by rsubtil** [comment]
There's another issue I didn't notice; the size of an `ItemImage` remains the same, so any size changes do not get updated. The `queue_redraw` is not enough, it needs some text changes to occur:

https://github.com/godotengine/godot/assets/6501975/433bbd88-8f07-4e56-93b0-4f8beddb881c

File: scene/gui/rich_text_label.cpp (hunk 1 above)
Comment (bruvzg) [reply scene/gui/rich_text_label.cpp]: This would make sense with the original version, but now it's doing more stuff.

**Comment by bruvzg** [comment]
Updated to fix size changes as well. Also replaced some nasty "cast raw item pointer to `uint64_t`" code with safer RIDs.

**Comment by rsubtil** [comment]
Can confirm it now works like a charm! :tada:

File: scene/gui/rich_text_label.cpp [hunk 2]
```diff
@@ -932,7 +933,8 @@ int RichTextLabel::_draw_line(ItemFrame *p_frame, int p_line, const Vector2 &p_o
 		// Draw inlined objects.
 		Array objects = TS->shaped_text_get_objects(rid);
 		for (int i = 0; i < objects.size(); i++) {
-			Item *it = reinterpret_cast<Item *>((uint64_t)objects[i]);
+			Item *it_span = items.get_or_null(objects[i]);
+			Item *it = reinterpret_cast<Item *>(it_span);
```
Comment (AThousandShips) [inline scene/gui/rich_text_label.cpp:937]: Unnecessary I think, just collapse to
```cpp
Item *it_span = items.get_or_null(objects[i]);
```

File: scene/gui/rich_text_label.cpp [hunk 3]
```diff
@@ -1601,7 +1603,8 @@ float RichTextLabel::_find_click_in_line(ItemFrame *p_frame, int p_line, const V
 
 		Array objects = TS->shaped_text_get_objects(rid);
 		for (int i = 0; i < objects.size(); i++) {
-			Item *it = reinterpret_cast<Item *>((uint64_t)objects[i]);
+			Item *it_span = items.get_or_null(objects[i]);
+			Item *it = reinterpret_cast<Item *>(it_span);
```
Comment (AThousandShips) [inline scene/gui/rich_text_label.cpp:1607]: Ditto

**Review by AThousandShips — APPROVED** [review]
Looks good otherwise

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**